### PR TITLE
Datasets: include the native CRS in each sample

### PR DESCRIPTION
**Pull request by adriantre**
### Summary
(follows torchgeo#3804, same motivation). 
GeoDatasets currently enforce a single CRS across the dataset; reprojection can be expensive, lossy, and conflicts with inference-time use where the native CRS is preferable. This PR makes samples self-describing so downstream code knows the CRS of the data it holds, setting up mixed-CRS batches and stitching of gridded predictions.

The effort to implement this is split into smaller PRs to make it easy to review. Node color shows each step's status relative to this PR: **green = implemented in this PR**, blue = already in place before it, gray = later in the stack (including deferred/blocked work). Data inputs are drawn as cylinders.

```mermaid
flowchart TD
  classDef prior fill:#d6eaf8,stroke:#2e86c1,color:#000
  classDef thisPR fill:#d5f5e3,stroke:#27ae60,color:#000
  classDef future fill:#eceff1,stroke:#b0bec5,color:#000

  FILES[("Raster files — native CRS & res<br/>(a dataset may span multiple zones)")]:::prior

  subgraph INIT["Dataset __init__"]
    IDX["index: geometry → index CRS (lookup only)<br/>+ native_crs, native_res per file"]:::future
    IDXCRS["index CRS = EPSG:6933 (global, equal-area)<br/>when prefer_native_crs"]:::future
    DEFAULT["prefer_native_crs default = False<br/>flip to native-by-default blocked by STITCH"]:::future
    REG["crs_registry: ordered CRSs a sample can emit<br/>(here just [self.crs]; frozen at init)"]:::thisPR
  end

  subgraph SMPL["Sampler (box in index CRS — no redesign)"]
    Q["query box in index CRS"]:::prior
  end

  subgraph GI["Standalone __getitem__"]
    LOOK["lookup df (.cx, index CRS)"]:::prior
    SEL["_select_out_crs — majority native CRS<br/>(warp the minority onto it)"]:::future
    REP["_reproject_slice: reproject CENTER point +<br/>box snapped to out_res grid (pixel-aligned, no resample)"]:::future
    READ["read in out_crs — NO warp when native"]:::future
    SMP["sample: data, bounds, transform,<br/>crs_index = crs_registry.index(out_crs)"]:::thisPR
  end

  subgraph CMB["Intersection / Union — multi-CRS combine"]
    VOTE["_resolve_out_crs: anchor = left-most child<br/>finest-GSD voting = override seam (future)"]:::future
    KEY["grid-tagged key (out_crs, out_res) on the GeoSlice;<br/>_split_grid peels it in each public __getitem__"]:::future
    CA["anchor child → native (no warp)"]:::future
    CB["other children (Raster / Vector / nested combo)<br/>→ warp / rasterize into anchor grid"]:::future
    SMPC["combined sample: combiner drops children's<br/>crs_index, stamps its own registry index"]:::thisPR
  end

  subgraph BATCH["DataLoader → Trainer"]
    COL["stack_samples / default_collate:<br/>crs_index is a tensor, stacks natively (no special-case)"]:::prior
    AUG["Kornia / device transfer: crs_index passes through untouched"]:::prior
  end

  subgraph INF["Inference"]
    PRED["per-chip predictions (crs_index in batch)"]:::prior
    STITCH["stitch gridded predictions via crs_registry[crs_index]<br/>→ reproject / mosaic to target CRS"]:::future
  end

  FILES --> IDX --> Q
  Q --> LOOK --> SEL
  SEL -->|"native != index"| REP --> READ
  SEL -->|"already aligned"| READ
  READ --> SMP
  REG --> SMP
  Q --> VOTE --> KEY
  KEY --> CA --> SMPC
  KEY --> CB --> SMPC
  SMP --> COL
  SMPC --> COL
  COL --> AUG --> PRED --> STITCH
```

### Additions

- Add a `crs_index` key to every GeoDataset sample: a 0-d integer tensor indexing a new dataset-side `crs_registry`, so `dataset.crs_registry[int(sample['crs_index'])]` is the sample's CRS.
- New `crs_registry` property on `GeoDataset`, frozen at `__init__`. Here it is `[self.crs]`; later PRs populate it with the distinct native CRSs.
- `IntersectionDataset`/`UnionDataset` stamp one coherent CRS on the combined sample: drop the children's `crs_index` (each points into its own registry) and re-stamp `self.crs`. This is the hook a later PR will use to resolve the combined CRS *from* the children, instead of defaulting to `self.crs`.

### Motivations

- Make samples self-describing so downstream code knows the CRS of the data it holds — the carrier for mixed-CRS batches and prediction stitching.
- Carry the CRS as an **integer index rather than a `pyproj.CRS` object**, so samples stay `dict[str, Tensor]`: PyTorch `default_collate`, `stack_samples`, and Lightning `transfer_batch_to_device` all work unchanged, no per-sample-list special-casing. The lossless `pyproj.CRS` lives once in `crs_registry`, avoiding WKT/EPSG-encoding loss (`to_epsg()` is `None` for `OGC:CRS84` and a wrong close-match for ESRI Albers).
- The registry is a deterministic, per-process function of the sorted files, so `crs_index` resolves to the same CRS in every independently built process (DataLoader workers, DDP ranks) with zero IPC. This holds trivially here (`[self.crs]`);. It gets a real test once the registry becomes multi-CRS.

### Behaviour / known limitations

- Behaviour is unchanged after this PR: `crs_registry` is `[self.crs]`, so every `crs_index` is `0` and resolves to the index CRS. Because it's sourced from `crs_registry.index(out_crs)`, it will automatically reflect any future per-query CRS selection without further wiring.
- Resolution requires the dataset (`crs_registry[crs_index]`), not the raw value. Named `crs_index` (not `crs`) so anyone calling `sample['crs']` fails loudly rather than returning a misleading int.

@adamjstewart @evgeniialappo @calebrob6 

### Checklist

- [X] I have read the [Contributing docs](https://docs.torchgeo.org/en/stable/user/contributing.html)
- [X] I have read the [AI policy](https://github.com/torchgeo/governance/blob/main/AI-POLICY.md)

- [ ] 🟢 **No AI usage**: written by humans, for humans
- [X] 🟡 **AI-assisted**: AI helped with the coding, but I understand every line
- [ ] 🔴 **AI-generated**: AI did everything, review with caution

### CODE REVIEW

**Comment by adamjstewart** [comment]
Was the PR description written by you or AI?

**Comment by adriantre** [comment]
> Was the PR description written by you or AI?

The graph is AI and revised by me. The text is me, revised by AI in an effort to bring clarity.

Let me know if anything is unclear.

**Comment by adamjstewart** [comment]
I ask because the description is extremely verbose and hard to follow. It's rare to see a description longer than the actual code. According to the AI policy, all descriptions should be written by humans and be as concise as possible.

**Comment by adriantre** [comment]
It is descriptive because it lays the foundation for future PRs. I need reviewers to understand more than just the code at hand. I need to onboard reviewers on the overall scope and direction while keeping each PR small. 

If it was under-descriptive I expect reviewers would say "this code does nothing for multi-CRS, it only adds complexity".

What parts of it do you consider hard to follow? Please help me help you

**Comment by calebrob6** [comment]
Common man it isn't hard to follow esp. if you review the previous PR:
- GeoDatastets now get a registry of CRSs that their samples might be in
- samples have an index into that registry
- its designed this way to not break the [str, Tensor] sample type
- we do this because we eventually want to have no reprojection as an option in mixed CRS datasets

it took longer to write this comment than parse. You don't have to make everything about AI and can just engage @adriantre in good faith. He's obviously not a bot and has thought a lot about this AND how to present this.
